pkg/parse/run.go

```diff
@@ -533,39 +531,46 @@ func (r *reconciler) setSyncStatus(ctx context.Context, newSyncStatus *SyncStatu
 	return nil
 }
 
-// updateSyncStatusPeriodically update the sync status periodically until the
-// cancellation function of the context is called.
-func (r *reconciler) updateSyncStatusPeriodically(ctx context.Context) {
-	opts := r.Options()
-	state := r.ReconcilerState()
-	klog.V(3).Info("Periodic sync status updates starting...")
-	updatePeriod := opts.StatusUpdatePeriod
-	updateTimer := opts.Clock.NewTimer(updatePeriod)
-	defer updateTimer.Stop()
-	for {
-		select {
-		case <-ctx.Done():
-			// ctx.Done() is closed when the cancellation function of the context is called.
-			klog.V(3).Info("Periodic sync status updates stopped")
-			return
-
-		case <-updateTimer.C():
-			klog.V(3).Info("Updating sync status (periodic while syncing)")
-			// Copy the spec and commit from the source status
-			syncStatus := &SyncStatus{
-				Spec:       state.status.SourceStatus.Spec,
-				Syncing:    true,
-				Commit:     state.cache.source.commit,
-				Errs:       r.ReconcilerState().SyncErrors(),
-				LastUpdate: nowMeta(opts.Clock),
+// startAsyncStatusUpdates starts a goroutine that updates the sync status
+// periodically until the cancel function is called.
+// Wait until the done channel is closed to confirm the goroutine has exited.
+func (r *reconciler) startAsyncStatusUpdates(ctx context.Context) (context.CancelFunc, <-chan struct{}) {
+	ctx, cancelFn := context.WithCancel(ctx)
```

Comment:
nit: I think this function could just take a context and return a doneCh, and let the caller handle creating the cancel function

Reply:
That's what it was doing before. I moved it inside the func to avoid the caller needing to manage two contexts, which seemed confusing.

Reply:
Yeah no biggie. I'm curious whether it's conventional to return the cancel function as the second parameter? It seems like it's normally the second return parameter, although I don't know if I've seen it with a doneCh as the other parameter

Reply:
The most idiomatic way I can think of would be to have a function that takes a Context and returns a doneCh. That's along the lines of what you were proposing. It just makes the caller do more work.

Reply:
refactored. used async prefix to distinguish the ctx.